packages/codecs-core/src/codec.ts

```diff
@@ -4,36 +4,63 @@
 export type Offset = number;
 
 /**
- * The shared attributes between codecs, encoders and decoders.
- */
-export type CodecData = {
-    /** A description for the codec. */
-    description: string;
-    /** The fixed size of the encoded value in bytes, or `null` if it is variable. */
-    fixedSize: number | null;
-    /** The maximum size an encoded value can be in bytes, or `null` if it is variable. */
-    maxSize: number | null;
+ * An object that can encode a value to a `Uint8Array`.
+ */
+export type Encoder<T> = EncoderSize<T> & {
+    /** Encode the provided value and return the encoded bytes directly. */
+    readonly encode: (value: T) => Uint8Array;
+    /**
+     * Writes the encoded value into the provided byte array at the given offset.
+     * Returns the offset of the next byte after the encoded value.
+     */
+    readonly write: (value: T, bytes: Uint8Array, offset: Offset) => Offset;
```

Comment:
You could achieve the context-y suggestion above by doing something like this:


```suggestion
    readonly write: (
        this: Readonly<{fixedSize: number | null}>,
        value: T,
        bytes: Uint8Array, offset: Offset,
    ) => Offset;
```

Then something like this later:

```ts
const context = {fixedSize};
return {
  // ...
  write(...args) { return encoder.write.apply(context, args); },
}
```